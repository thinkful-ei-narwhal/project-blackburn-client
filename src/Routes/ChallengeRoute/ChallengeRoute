import React, { Component } from "react";
import "./ChallengeRoute.module.css";

class ChallengeRoute extends Component {
  componentDidMount() {}

  render() {
    return (
      <div>
        <p>TESTING</p>
      </div>
    );
  }
}

export default ChallengeRoute;
