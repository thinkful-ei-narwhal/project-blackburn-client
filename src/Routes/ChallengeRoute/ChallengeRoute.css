.game-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  border: 1px solid grey;
  width: auto;
  /* background-color: #d3d3d3; */
  /* background-image: url('../../Images/DetectiveAtDesk.jpg'); */
  background-size: cover;
  position: absolute;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
}

.challenge-div,
.winlose-div {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.word-ul {
  /* border: 1px solid grey; */
  padding: 10px;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  width: 100%;
  height: 30em;
  background-color: lightgrey;
  box-shadow: 3px 3px 3px 3px black;
}

.bg {
  background-color: black;
  height: 40px;
}

.word-li {
  /* margin: 10px */
}

/* .wordTimer {
    float: left;
} */

.time-ui {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: 100%;
  border: 1px solid gray;
  box-shadow: 4px 4px 4px 4px lightgray;
  /* padding: 20px; */
  /* margin: 15px; */
}

.stats-container {
  display: flex;
  flex-direction: row;
  justify-content: center;
  justify-self: center;
  align-items: center;
  align-self: center;
  text-align: center;
  border: 1px solid gray;
  max-width: 46vh;
  box-shadow: 4px 4px 4px 4px black;
  background-color: lightgrey;
  font-size: larger;
}

.stat {
  text-align: center;
  margin: 10px;
  border-bottom: 1px solid lightgray;
}

.type-input {
  margin: 5px;
  text-align: center;
}

.word-timer {
  border-radius: 50px;
  background-color: black;
  /* width: 50px; */
}


.hint {
  margin: 10px;
  font-size: 2.5vh;

}

.hint-container {
  position: fixed;
  top: 50vh;
  bottom: 50vh;
  background-color: rgba(100, 0, 100, .4);
  height: 20vh;
  width: 30vh;
  border-radius: 10px;
  color: white;
  font-size: 2.5vh;
}