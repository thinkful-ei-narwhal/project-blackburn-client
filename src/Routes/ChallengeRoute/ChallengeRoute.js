import React, { Component } from "react";
import TypeHandler from "./../../Components/TypeHandler/TypeHandler";
import Healthbar from "./../../Components/Healthbar/Healthbar";
import Score from "./../../Components/Score/Score";
import Wpm from "./../../Components/Wpm/Wpm";
import Word from "./../../Components/Word/Word";
import GameplayScreen from "./../../Components/GameplayScreen/GameplayScreen";
import ScoreboardApiService from "./../../Services/scoreboard-api-service";
import { uniqueNamesGenerator, animals } from "unique-names-generator";
import "./ChallengeRoute.module.css";

class ChallengeRoute extends Component {
  state = {
    words: [],
    playerHealth: 5,
    playerScore: 0,
    playerBest: 0,
    playerBestStored: 0,
    typedWords: 0,
    // wpmAvgs: [],
    // wpmCalc: 0,
    levelTimer: 60, //1 minute
    levelTimerTotal: 120,
    initialized: false,
  };

  //smaller Todos:
  //Research how to make word components appear at different places on the screen
  //screen shake > part of typeHandler, if it's correct or incorrect
  //need to get timers for each word appearing

  //Todo get the WPM tracker working
  // pushWpmValue(state) {
  //   let typedWords = state.typedWords;
  //   const wpmAvgs = state.wpmAvgs;
  //   const currentWPM = typedWords / (state.levelTimerTotal / 60);
  //   if (currentWPM !== wpmAvgs[wpmAvgs.length - 1]) {
  //     wpmAvgs.push(currentWPM);
  //   }
  //   else {
  //     wpmAvgs.push(0);
  //   }

  //   let totalWPM = 0;
  //   for (let i = 0; i < wpmAvgs.length; i++) {
  //     totalWPM += wpmAvgs[i];
  //   }
  //   const wpmCalc = totalWPM / wpmAvgs.length;

  //   this.setState({ wpmAvgs, wpmCalc });
  // }

  // updateLevelTimer() {
  //   let levelTimer = this.state.levelTimer;
  //   levelTimer--;
  //   this.setState({ levelTimer });
  // }

  manageWords(addWordObj = null) {
    //removes all words that are expired
    const newWordArray = this.state.words.filter(
      (wordObj) => wordObj.expired === false
    );

    //adds a new word if it's updated
    if (addWordObj !== null) newWordArray.push(addWordObj);
    this.setState({ words: newWordArray });
  }

  generateWord(wordTimeout, maxWords) {
    if (this.state.words.length < maxWords) {
      const randomWord = uniqueNamesGenerator({
        dictionaries: [animals],
        length: 1,
      });

      const newWord = {
        word: randomWord,
        expired: false,
        timeout: setTimeout(() => {
          newWord.expired = true;
          this.setState({ playerHealth: this.state.playerHealth - 0.5 });
          this.manageWords();
        }, wordTimeout),
      };

      this.manageWords(newWord);
    }
  }

  handleSubmit(event, state) {
    event.preventDefault();
    const userInput = event.target.typeInput.value;
    event.target.typeInput.value = "";

    const newWords = state.words;
    let playerHealth = state.playerHealth;
    let playerScore = state.playerScore;
    let playerBest = state.playerBest;
    let playerBestStored = state.playerBestStored;
    let typedWords = state.typedWords;

    let takeDamage = true;
    newWords.forEach((wordObj) => {
      if (wordObj.word.toLowerCase() === userInput.toLowerCase()) {
        takeDamage = false;
        wordObj.expired = true;
        playerScore += 10;
        typedWords++;
        clearTimeout(wordObj.timeout);
      }
      return;
    });

    //if the player is incorrect he takes damage and losses score
    if (takeDamage) playerHealth--;
    if (takeDamage) playerScore -= 5;

    //handle player record calculation
    if (playerBest <= playerScore) playerBest = playerScore; //css should turn green when it surpasses old score
    if (playerBest > playerScore) playerBest = playerBestStored; //css should return to black if it dips back below old score

    this.setState({
      words: newWords,
      playerHealth,
      playerScore,
      playerBest,
      typedWords,
    });
    this.manageWords();
  }

  componentDidMount() {
    //todo get all of the data you need and set contexts!
    // ScoreboardApiService.getScoreboard(3,"story",1).then(res => {
    //   console.log('TESTING ', res);
    // });
    //Values this route will eventually need to get on start:
    //So in componentDidMount we need:
    //>>story, difficulty, player's high score

    this.levelTimeout = setInterval(() => this.updateLevelTimer(), 1000);
    this.intervalGenerator = setInterval(
      () => this.generateWord(20000, 5),
      2000
    );
    this.wpmAvgCalculator = setInterval(
      () => this.pushWpmValue(this.state),
      1000
    );
    this.setState({ initialized: true });
  }

  componentWillUnmount() {
    clearInterval(this.intervalGenerator);
    clearInterval(this.levelTimeout);
    clearInterval(this.wpmAvgCalculator);
  }

  renderGameplay() {
    return (
      <div>
        <p>
          Time Remaining:{" "}
          {this.state.levelTimer >= 0 ? this.state.levelTimer : 0}
        </p>
        <Healthbar health={this.state.playerHealth} />
        <div>
          <span>Personal best: </span>
          <Score score={this.state.playerBest} />
        </div>
        <div>
          <span>Score: </span>
          <Score score={this.state.playerScore} />
        </div>
        <span>Words Per Minute: </span>
        <Wpm wpm={this.state.wpmCalc} />
        <TypeHandler handleSubmit={(e) => this.handleSubmit(e, this.state)} />
        <ul>
          {this.state.words.map((wordObj, index) => (
            <li key={index}>
              <Word word={wordObj.word} />
            </li>
          ))}
        </ul>
        <GameplayScreen />
      </div>
    );
  }

  render() {
    return this.state.initialized ? this.renderGameplay() : null;
  }
}

export default ChallengeRoute;
