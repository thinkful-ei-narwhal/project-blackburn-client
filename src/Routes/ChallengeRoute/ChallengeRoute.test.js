import React from 'react';
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import BlackBurnContext from '../../Context/BlackburnContext';
import ChallengeRoute from './ChallengeRoute';

describe(`Renders challenge page to dom without crashing`, () => {
    it('renders checkpoint win crashing', () => {
        const wrapper = shallow(
        <BlackBurnContext.Provider>
            <ChallengeRoute />
        </BlackBurnContext.Provider>
    )
    .find('.game-container')
    expect(toJson(wrapper)).toMatchSnapshot()
    })
})