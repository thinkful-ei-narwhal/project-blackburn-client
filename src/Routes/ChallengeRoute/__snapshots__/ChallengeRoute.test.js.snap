// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders challenge page to dom without crashing renders checkpoint win crashing 1`] = `null`;
